design-docs/incremental-build.md

```diff
@@ -243,3 +243,109 @@ When the `TaskHistory` gets persisted, it adds the current task execution to the
+
+The goal of this story is to 1) detect this situation and 2) clean up any stale output.
+
+### User visible changes
+
+Gradle detects and removes stale Java class files after a Gradle version change. Further processing of output files (e.g. by the `jar` task) does not result in faulty behavior.
+
+### Implementation
+
+- Establish a registry implementation that allows for registering strategies with the purpose of cleaning up outputs produced by previous builds.
+    - The registry is not specific to tasks. It's rather a global concept that allows for registering clean up strategies for any "thing" in Gradle.
+    - The registry is going to be an internal concept and is not going to be exposed via a public API.
+    - The registry needs to be instantiated by Gradle's global service registry. 
+    - When a build is executed, the registry is contacted and provides all registered strategies.
+    - If a strategy indicates that clean up is needed then delete all registered outputs.
+
+<!-- -->
+
+    public interface BuildOutputCleanupStrategy {
+        /**
+         * Determines if cleanup is needed based on provided Gradle instance.
+         */
+        boolean needsCleanup(Gradle gradle);
+    }
+    
+    public interface BuildOutputCleanup {
+        /**
+         * Returns the clean up strategy.
+         */
+        BuildOutputCleanupStrategy getStrategy();
+        
+        /**
+         * Returns the outputs that need to be cleaned if needed.
+         */
+        List<File> getOutputs();
+    }
+
+    public interface BuildOutputCleanupRegistry {
+        /**
+         * Registers clean up strategy for given outputs.
+         */
+        void registerCleanup(BuildOutputCleanup cleanups);
+        
+        /**
+         * Returns all clean ups.
+         */
+        List<BuildOutputCleanup> getCleanups();
+    }
+
+- Provide an implementation for a clean up strategy that identifies if task history is available.
+    - Create an instance of the implementation in Gradle's global service registry.
+    - Check if the `.gradle/<version>` directory exists for current version of Gradle used for build.
+    - Stale files produced by older versions of Gradle (that do not know about the registry implementation) are out of scope.
+
+<!-- -->
+
+    public class TaskHistoryCleanupStrategy implements BuildOutputCleanupStrategy {
+        @Override
+        public boolean needsCleanup(Gradle gradle) {
+            // identify existing task history
+        }
+    }
+
+- Clean up registry concept applies to output of Java compilation.
+    - Register the task history clean up implementation with the clean up registry in the `JavaPlugin`. 
+    - As output define outputs of the `main` and `test` source set.
+    - Add classes output directory via `sourceSet.getOutput().getClassesDir()`.
+    - Add resources output directory via `sourceSet.getOutput().getResourcesDir()`.
```

Comment:
Do you have a good idea how to deal with the situation? Maybe we could just defer cleaning the directories to the execution phase.